src/controllers/helpers/pagination.rs

```diff
@@ -215,59 +216,88 @@ where
-        let page_error = Page::new(&mut req).unwrap_err().response().unwrap();
+        assert_error("page=", "cannot parse integer from empty string");
+        assert_error("page=not_a_number", "invalid digit found in string");
+        assert_error("page=1.0", "invalid digit found in string");
+        assert_error("page=0", "page indexing starts from 1, page 0 is invalid");
 
-        assert_eq!(page_error.status(), StatusCode::BAD_REQUEST);
+        let pagination = PaginationOptions::builder()
+            .gather(&mut mock("page=5"))
+            .unwrap();
+        assert_eq!(Page::Numeric(5), pagination.page);
     }
 
     #[test]
-    fn page_must_be_non_zero_positive_integer() {
-        let mut req = mock("page=0");
-        let page_error = Page::new(&mut req).unwrap_err().response().unwrap();
+    fn per_page_param_parsing() {
+        let assert_error =
+            |query, msg| assert_pagination_error(PaginationOptions::builder(), query, msg);
 
-        assert_eq!(page_error.status(), StatusCode::BAD_REQUEST);
+        assert_error("per_page=", "cannot parse integer from empty string");
+        assert_error("per_page=not_a_number", "invalid digit found in string");
+        assert_error("per_page=1.0", "invalid digit found in string");
+        assert_error("per_page=101", "cannot request more than 100 items");
 
-        let mut req = mock("page=1.0");
-        let page_error = Page::new(&mut req).unwrap_err().response().unwrap();
+        let pagination = PaginationOptions::builder()
+            .gather(&mut mock("page=5"))
+            .unwrap();
+        assert_eq!(Page::Numeric(5), pagination.page);
+    }
 
-        assert_eq!(page_error.status(), StatusCode::BAD_REQUEST);
+    fn mock(query: &str) -> MockRequest {
+        let mut req = MockRequest::new(http::Method::GET, "");
+        req.with_query(query);
+        req
     }
 
-    #[test]
-    fn per_page_must_be_a_number() {
-        let mut req = mock("per_page=");
-        let per_page_error = PaginationOptions::builder()
-            .gather(&mut req)
+    fn assert_pagination_error(options: PaginationOptionsBuilder, query: &str, message: &str) {
+        let response = options
+            .gather(&mut mock(query))
             .unwrap_err()
             .response()
             .unwrap();
-        assert_eq!(per_page_error.status(), StatusCode::BAD_REQUEST);
 
-        let mut req = mock("per_page=not_a_number");
-        let per_page_error = PaginationOptions::builder()
-            .gather(&mut req)
-            .unwrap_err()
-            .response()
-            .unwrap();
-        assert_eq!(per_page_error.status(), StatusCode::BAD_REQUEST);
+        assert_eq!(StatusCode::BAD_REQUEST, response.status());
+
+        // Conduit doesn't expose a method to get the contents of a body, so we need to extract it
+        // from the underlying body variants. Please replace this with a proper method if conduit
+        // ever adds support for that.
```

Comment:
It's not very discoverable, but you can `use conduit_test::ResponseExt;` and then call `request.into_cow()` which returns a `Cow<[u8]>`.

Reply:
TIL!